### DISCUSSION TITLE
Use of normalized (e.g. with hyphens) name in console scripts

### DISCUSSION THREAD

**Discussion by ChrisBarker-NOAA** [pypa/setuptools#3459] [General]
I recently happened upon a weird bug whereby my console scripts failed:
(see: https://discuss.python.org/t/whats-up-with-scripts-and-hyphens-underscores/17571)

It turns out my issue was due to a mess of broken interactions between conda environments, editable installs, and multiple git branches -- so something got weirdly out of sync -- all my fault.

However, in the process, I noticed that the console script wrapper is using the hyphenated name for teh package, which seems really odd to me.

```
__requires__ = 'adios-db'
...
    sys.exit(load_entry_point('adios-db', 'console_scripts', 'adios_db_process_json')())
```

The Python package name is `adios_db` and I don't specify a hyphenated version anywhere.

I understand the distinction between a "package" and a "distribution", and that a distribution does not need to have a name that is a Python legal identifier. And that a distribution name of "adios_db" and "adios-db" are considered the same, and that the concept of a normalized name is important.

( I have no idea why the normalized form is with the hyphen, rather than the underscore, but I'll leave that alone for now)

What I don't get is why the normalized distribution name is used in this context, rather than the python package name -- it seems to me that the package name is the core, "proper" name when used in python code -- and that a console script should be identified with the python package, not the distribution.

and I'm very confused as to what would happen if the distribution provided multiple packages and / or a package with a different name than the distribution. After all, I made no mention of the distribution when I set up the console script:

```
scripts = [...
           'adios_db_process_json = adios_db.scripts.process_json:run_through',
           ...
           ]
```

That seems very clearly to mean:

the script called `adios_db_process_json` should call the `run_through` function in the `adios_db.scripts.process_json` module.

There is no statement there about what distribution this may be a part of.

I guess this all reflects my general discomfort with pushing the concept of a "distribution"  into Python itself, at run time, rather than keeping python itself all about packages and modules, and the distribution concept being about, well, the distribution of packages, and no longer reevent at run time.

And it seems gratuitous -- once you are running a console script, the code that needs to be run has to be installed as a proper python package, accessible via the usual import mechanism anyway -- shouldn't the question of what distribution it was installed as part of be irrelevant?

I suppose this is all philosophical -- it works as it is -- but I'd still like to know why it's done this way -- it feels like the wrong direction to be heading to me.

**Comment by abravalheri** [comment]
Hi @ChrisBarker-NOAA , I have a question, how are you generating the script wrappers?

Are you installing the project via `pip install .` or are you using `pip install -e .`? Does using the 2 different methods affect the generated wrappers?

**Comment by ChrisBarker-NOAA** [comment]
Bingo! -- I had been using:

`pip install -e` 

And when I do `pip install ./ `, e.g regular install, I get a far simpler and more straightforward wrapper script:

```
import re
import sys
from adios_db.scripts.process_json import run_through
if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
    sys.exit(run_through())
```
why I didn't think to check that myself, I don't know -- sorry for the noise.

I do see why an editable install would be more complex, though I'm still confused about the use of the normalized distribution name, rather than the regular module name -- but if it's only for the editable installs -- then it won't leak out in a risky way.

(and now that I think about it -- why wouldn't this simple approach work for editable installs as well?)

Thanks for the quick reply,

-CHB

**Reply by abravalheri** [reply]
We are in the process of updating the editable install (PEP 660), so probably once that is implemented, both wrappers should be the same (the simpler one).

**Reply by ChrisBarker-NOAA** [reply]
sounds great -- thanks!
